In [1]:
import numpy as np
import pandas as pd

## 项目: 整理 Netflix 电影演员评分数据

#### 分析目标

此数据分析的目的是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，歌演员出演作品的平均 IMDB 评分，从而挖掘出各个流派中的高评分作品演员。<br>
<br>
本实战项目的目的在于练习整理数据，从而得到可供下一步分析的数据。

#### 简介

原始数据记录了截止至 2022年7月 美国地区可观看的所有 Netflix 电视剧及电影数据。数据集包含两个数据表: `titles.csv`和`credits.csv`<br>
<br>
`titles.csv`包含电影及电视剧相关信息，包括影视作品ID、标题、类型、描述、流派、IMDB (一个国外的在线评分网站)评分，等等。<br>
`credits.csv`包含超过7万名出现在 Netflix 影视作品的导演及演员信息，包括名字、影视作品ID、人物名、演职员类型 (导演/演员)等。<br>

`titles.csv`每列的含义如下:<br>
> - id: 影视作品ID。
> - title: 影视作品标题。
> - show_type: 作品类型，电视节目或电影。
> - description: 简短描述。
> - release_year: 发布年份。
> - age_certification: 适龄认证。
> - runtime: 每集电视剧或电影的长度。
> - genres: 流派类型列表。
> - production_countries: 出品国家列表。
> - seasons: 如果是电视剧则是季数。
> - imdb_id: IMDB 的 ID。
> - imdb_score: IMDB 的投票数。
> - imdb_votes: IMDB 的投票数。
> - tmdb_popularity: TMDB 的流行度。
> - tmdb_score: TMDB 的评分。 

`credits.csv`每列的含义如下:
> - person_ID: 演职员 ID。
> - id: 参与的影视作品 ID。
> - name: 姓名。
> - character_name: 角色姓名。
> - role: 演职员类型演员或导演。

#### 读取数据

导入数据分析所需要的库，并通过 Pandas 的`read_csv`函数，<br>
将原始数据文件`titles.csv`里的数据内容，解析为DataFrame并赋值给变量`Original_Titles`。<br>
将原始数据文件`credits.csv`里的数据内容，解析为 DataFrame 并赋值给变量`Original_Credits`。

In [2]:
Original_Titles = pd.read_csv(".//Organizing_Data_Practice_From_Netflix//titles.csv")
Original_Credits = pd.read_csv(".//Organizing_Data_Practice_From_Netflix//credits.csv")

In [3]:
Original_Titles.head()

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600

In [4]:
Original_Credits.head()

person_id       id             name                character   role
0       3748  tm84618   Robert De Niro            Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster            Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                      Tom  ACTOR
3       3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                    Betsy  ACTOR

#### 评估和清理数据

在这一部分中，我们将对在上一部分建立的`Original_Titles`及`Original_Credits` DataFrame 所包含的数据进行评估和清理。<br>
<br>
主要从两个方面内容: 结构和内容，即整齐度和干净度。<br>
<br>
数据的结构性问题指不符合"每个变量为一列，每个观察值为一行，每个类型的观察单位为一个表格"这三个标准; 数据的内容性问题包括存在丢失数据、重复数据、无效数据等。<br>
<br>
为了区分开经过清理的数据和原始的数据，我们创建新的变量`Cleaned_Titles`, 让它为`Original_Titles`复制出的副本，以及创建新的变量`Cleaned_Credits`, 让它为`Original_Credits`复制出的变量。我们之后的清理步骤都将被运用在`Cleaned_Titles`和`Cleaned_Credits`上。

In [5]:
Cleaned_Titles = Original_Titles.copy()
Cleaned_Credits = Original_Credits.copy()

#### 数据整齐度

In [6]:
Cleaned_Titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   
5   ts22164         Monty Python's Flying Circus   SHOW   
6   tm70993                        Life of Brian  MOVIE   
7   tm14873                          Dirty Harry  MOVIE   
8  tm119281                     Bonnie and Clyde  MOVIE   
9   tm98978                      The Blue Lagoon  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   
5  A British sketch comedy series with the shows ...          1969   
6  Brian Cohen is an average young Jewish man, bu...          1979   
7  When a madman dubbed 'Scorpio' terrorizes San ...          1971   
8  In the 1930s, bored waitress Bonnie Parker fal...          1967   
9  Two small children and a ship's cook survive a...          1980   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   
5             TV-14       30                       ['comedy', 'european']   
6                 R       94                                   ['comedy']   
7                 R      102              ['thriller', 'action', 'crime']   
8                 R      110                 ['crime', 'drama', 'action']   
9                 R      104               ['romance', 'action', 'drama']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   
5               ['GB']      4.0  tt0063929         8.8     73424.0   
6               ['GB']      NaN  tt0079470         8.0    395024.0   
7               ['US']      NaN  tt0066999         7.7    155051.0   
8               ['US']      NaN  tt0061418         7.7    112048.0   
9               ['US']      NaN  tt0080453         5.8     69844.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600  
5           17.617       8.306  
6           17.770       7.800  
7           12.817       7.500  
8           15.687       7.500  
9           50.324       6.156

从数据的部分 10 行来看, `Cleaned_Titles`里的`genres`和`production_countries`的变量包含多个值，应当进行拆分。<br>
<br>
先提取任意一个`genres`变量的值进行观察。

In [7]:
Cleaned_Titles['genres'][1]

"['drama', 'crime']"

虽然`genres`表示形式是列表, 但其实际类型并非字符串表，而是字符串，无法直接用`value_counts`统计各个值出现的次数，我们可以使用 Python 内置的`eval`函数, 它可以把字符串转换成表达式，所以可以帮助我们把表示列表的字符串转换成列表本身。

In [8]:
Cleaned_Titles['genres'] = Cleaned_Titles['genres'].apply(lambda s:eval(s))
Cleaned_Titles['genres'][1]

['drama', 'crime']

转换为列表后，就能用 DataFrame 的`explode`方法，把那个列的列表值拆分成单独的行。

In [9]:
Cleaned_Titles = Cleaned_Titles.explode('genres')
Cleaned_Titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation               ['US']      1.0   
1                 R      114          drama               ['US']      NaN   
1                 R      114          crime               ['US']      NaN   
2                 R      109          drama               ['US']      NaN   
2                 R      109         action               ['US']      NaN   
2                 R      109       thriller               ['US']      NaN   
2                 R      109       european               ['US']      NaN   
3                PG       91        fantasy               ['GB']      NaN   
3                PG       91         action               ['GB']      NaN   
3                PG       91         comedy               ['GB']      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

接下来，针对`production_countries`列也是一样的流程。<br>
每个观察值的`production_countries`值并不表示单个流派，而是一系列的流派。先提取任意一个`production_countries`变量的值进行观察。

In [10]:
Cleaned_Titles['production_countries'][1]

1    ['US']
1    ['US']
Name: production_countries, dtype: object

可以看到，`production_countries`也是一样的问题，虽然表示形式是列表，但其实际类型并非字符串表，而是字符串，难以进行拆分。我们可以再次利用`eval`函数进行类型转换，并检查转换后确实是列表类型。

In [11]:
Cleaned_Titles['production_countries'] = Cleaned_Titles['production_countries'].apply(lambda s:eval(s))
Cleaned_Titles['production_countries'][0]

['US']

确认类型转换完毕后，还是用`explode`方法，把列表拆分成单独的行。

In [12]:
Cleaned_Titles = Cleaned_Titles.explode('production_countries')
Cleaned_Titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation                   US      1.0   
1                 R      114          drama                   US      NaN   
1                 R      114          crime                   US      NaN   
2                 R      109          drama                   US      NaN   
2                 R      109         action                   US      NaN   
2                 R      109       thriller                   US      NaN   
2                 R      109       european                   US      NaN   
3                PG       91        fantasy                   GB      NaN   
3                PG       91         action                   GB      NaN   
3                PG       91         comedy                   GB      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

在处理完`Cleaned_Titles`的结构性问题后，查看`Cleaned_Credits`。

In [13]:
Cleaned_Credits.head(10)

person_id       id             name                      character   role
0       3748  tm84618   Robert De Niro                  Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster                  Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                            Tom  ACTOR
3       3739  tm84618    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                          Betsy  ACTOR
5      32267  tm84618      Peter Boyle                         Wizard  ACTOR
6     519612  tm84618   Leonard Harris      Senator Charles Palantine  ACTOR
7      29068  tm84618   Diahnne Abbott                Concession Girl  ACTOR
8     519613  tm84618      Gino Ardito             Policeman at Rally  ACTOR
9       3308  tm84618  Martin Scorsese  Passenger Watching Silhouette  ACTOR

从头部的 10 行来看，`Cleaned_Credits`数据符合"每个变量为一列，每个观察值为一行，每种类型的观察单位为一个表格"，因此不存在结构性问题。

#### 数据干净度

接下来通过`info`对数据内容进行大致了解。

In [14]:
Cleaned_Titles.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17818 entries, 0 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    17818 non-null  object 
 1   title                 17817 non-null  object 
 2   type                  17818 non-null  object 
 3   description           17790 non-null  object 
 4   release_year          17818 non-null  int64  
 5   age_certification     10889 non-null  object 
 6   runtime               17818 non-null  int64  
 7   genres                17755 non-null  object 
 8   production_countries  17439 non-null  object 
 9   seasons               6224 non-null   float64
 10  imdb_id               17116 non-null  object 
 11  imdb_score            16976 non-null  float64
 12  imdb_votes            16945 non-null  float64
 13  tmdb_popularity       17663 non-null  float64
 14  tmdb_score            17241 non-null  float64
dtypes: float64(5), int64(2), 

从输出结果来看, `Cleaned_Titles`数据共有 5850 条观察值, `title`、`description`、`age_certification`、`seasons`、`imdb_id`、`imdb_score`、`tmdb_popularity`、`tmdb_score`变量存在缺失值，将在后续进行评估和清理。<br>
<br>
此外, `release_year`表示年份，数据类型不应为数字，应为日期，所以需要进行数据格式转换。

In [15]:
Cleaned_Titles['release_year'] = pd.to_datetime(Cleaned_Titles['release_year'], format='%Y')
Cleaned_Titles['release_year']

0      1945-01-01
1      1976-01-01
1      1976-01-01
2      1972-01-01
2      1972-01-01
          ...    
5847   2021-01-01
5848   2021-01-01
5849   2021-01-01
5849   2021-01-01
5849   2021-01-01
Name: release_year, Length: 17818, dtype: datetime64[ns]

In [16]:
Cleaned_Credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  int64 
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.0+ MB


从输出结果来看，`Cleaned_Credits`数据共有 77801 条观察值，其中`character`变量存在缺失值，将在后续进行评估和清理。<br>
<br>
此外, `person_id`表示演职员 ID , 数据类型不应位数字, 应为字符串, 所以需要进行数据格式转换。

In [17]:
Cleaned_Credits['person_id'] = Cleaned_Credits['person_id'].astype('str')
Cleaned_Credits['person_id']

0           3748
1          14658
2           7064
3           3739
4          48933
          ...   
77796     736339
77797     399499
77798     373198
77799     378132
77800    1950416
Name: person_id, Length: 77801, dtype: object

#### 处理缺失数据

在`Cleaned_Title`中, `title`、`description`、`age_certification`、`seasons`、`imdb_id`、`imdb_score`、`tmdb_popularity`、`tmdb_score`变量存在缺失值。<br>
<br>
由于影视作品的标题、描述、适龄认证、电视剧季数、IMDB的ID、IMDB的流行度、TMDB的评分，并不影响我们挖掘各个流派中的高IMDB评分作品演员, 所以可以保留`title`、`description`、`age_certification`、`seasons`、`imdb_id`、`tmdb_popularity`、`tmdb_score`变量值存在空缺的观察值。<br>
<br>
但`imdb_score`，即IMDB评分，和我们后需要做的分析息息相关，所以提取出`imdb_score`缺失值进行查看。

In [18]:
Cleaned_Titles.query("imdb_score.isnull()")

id                                      title   type  \
0      ts300399        Five Came Back: The Reference Films   SHOW   
75     tm132164                       Bill Hicks: Sane Man  MOVIE   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
...         ...                                        ...    ...   
5810  tm1225897                                 Social Man  MOVIE   
5833   ts307884                                 HQ Barbers   SHOW   
5840  tm1216735                            Sun of the Soil  MOVIE   
5844  tm1074617              Bling Empire - The Afterparty  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   

                                            description release_year  \
0     This collection includes 12 World War II-era p...   1945-01-01   
75    Sane Man was filmed before Bill recorded ‘Dang...   1989-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
...                                                 ...          ...   
5810  Two competitive social media Influencers go he...   2021-01-01   
5833  When a family run barber shop in the heart of ...   2021-01-01   
5840  In 14th-century Mali, an ambitious young royal...   2022-01-01   
5844  The stars of "Bling Empire" discuss the show's...   2021-01-01   
5848  Jamie Foxx, David Alan Grier and more from the...   2021-01-01   

     age_certification  runtime         genres production_countries  seasons  \
0                TV-MA       51  documentation                   US      1.0   
75                   R       80         comedy                   US      NaN   
145               TV-G       18  documentation                   JP     12.0   
145               TV-G       18         family                   JP     12.0   
145               TV-G       18        reality                   JP     12.0   
...                ...      ...            ...                  ...      ...   
5810               NaN       96          drama                  NaN      NaN   
5833             TV-14       24         comedy                   NG      1.0   
5840               NaN       26            NaN                  NaN      NaN   
5844               NaN       35            NaN                   US      NaN   
5848             PG-13       37            NaN                   US      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0            NaN         NaN         NaN            0.600         NaN  
75           NaN         NaN         NaN            3.377         7.5  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
...          ...         ...         ...              ...         ...  
5810  tt20198164         NaN         NaN              NaN         NaN  
5833         NaN         NaN         NaN            0.840         NaN  
5840         NaN         NaN         NaN            1.179         7.0  
5844         NaN         NaN         NaN              NaN         NaN  
5848         NaN         NaN         NaN            1.296        10.0  

[842 rows x 15 columns]

由于缺失分析所需要的核心数据`imdb_score`, 我们将把这些观察值删除，并查看删除后该列空缺值个数和:

In [19]:
Cleaned_Titles = Cleaned_Titles.dropna(subset = ["imdb_score"])
Cleaned_Titles["imdb_score"].isnull().sum()

np.int64(0)

接下来评估`Cleaned_Credits`的缺失数据，其中只有`character`变量存在缺失值。<br>
<br>
角色名并不影响我们挖掘各个流派中的高IMDB评分作品演员, 并且此变量缺失也有可能因为演职员类别是导演, 没有对应角色, 因此可以保留`character`变量值存在空缺的观察值。

#### 处理重复数据

根据数据变量的含义以及内容来看, `Cleaned_Titles`里不应该存在每个变量值都相同的观察值, 因此查看是否存在重复值。

In [20]:
Cleaned_Titles.duplicated().sum()

np.int64(0)

输出结果为 0 , 说明不存在重复值。<br>
<br>
接下来看看`Cleaned_Credit`数据表是否存在重复值。

In [21]:
Cleaned_Credits.duplicated().sum()

np.int64(0)

输出结果为 0, 说明不存在重复值。

#### 处理不一致数据

针对`Cleaned_Titles`, 不一致数据可能存在于`genres`和`character`变量中，我们将查看是否存在多个不同值指代同一流派，以及多个不同值指代同一国家的情况。

In [22]:
Cleaned_Titles['genres'].value_counts()

genres
drama            3357
comedy           2419
thriller         1446
action           1339
romance          1080
crime            1066
documentation     981
family            769
animation         732
fantasy           727
european          679
scifi             647
horror            438
history           336
music             266
reality           226
war               221
sport             188
western            53
Name: count, dtype: int64

从上面看出, `genres`列里并不存在不一致数据，各个值都在指代不同的流派。但是里面还存在空字符串表示的流派，并非有效数据，因此可以进行删除。<br>
<br>
删除后，查看`Cleaned_Titles`里面是否还存在`genres`为空字符串的行:

In [23]:
Cleaned_Titles = Cleaned_Titles.query('genres != ""')
Cleaned_Titles.query('genres == ""')

Empty DataFrame
Columns: [id, title, type, description, release_year, age_certification, runtime, genres, production_countries, seasons, imdb_id, imdb_score, imdb_votes, tmdb_popularity, tmdb_score]
Index: []

接下来，针对`production_countries`列也是一样的流程, 利用`production_countries`的列表里面各个值的出现次数。

In [24]:
Cleaned_Titles['production_countries'].value_counts()

production_countries
US    5651
IN    1610
GB    1068
JP    1046
FR     720
      ... 
CU       1
LK       1
GT       1
AF       1
FO       1
Name: count, Length: 108, dtype: int64

由于`value_counts`执行结果中有太多值, Pandas只会默认显示开头和结尾的一些值，要完整显示结果，可以把`display.max_rows`设置为`None`, 即取消展示行上限。<br>
<br>
但因为我们只是在当前调用`value_counts`时才需要看完整结果，所以可以结合`option_content`, 只需改临时上限。

In [25]:
with pd.option_context('display.max_rows', None):
    print(Cleaned_Titles['production_countries'].value_counts())

production_countries
US         5651
IN         1610
GB         1068
JP         1046
FR          720
ES          637
KR          637
CA          608
DE          383
CN          295
MX          264
IT          224
BR          221
AU          217
TR          195
PH          192
AR          150
ID          149
BE          148
TW          133
NG          131
PL          126
ZA          103
HK          102
NL          102
CO           94
EG           93
DK           89
TH           87
SE           81
LB           70
NO           68
AE           52
IE           49
SG           47
XX           43
IL           42
RU           41
CL           35
CH           33
PS           32
BG           31
MY           30
SA           28
IS           28
AT           28
NZ           27
LU           27
PE           26
RO           25
QA           24
CZ           22
JO           19
FI           18
HU           18
UY           15
MA           15
PT           14
KH           10
KW           10
PR            9
PK 

从以上输出结果来看，出品国家都用两位的国家代码来表示，除了里面存在一个`Lebanon`值。<br>
<br>
`Lebanon`的国家代码是`LB`, 出现了39次，说明此处数据不一致。`LB`和`Lebanon`都在表示同一国家，需要进行统一。<br>
<br>
把`Cleaned_Titles`里，`production_countries`的`LB`和`Lebanon`统一为`LB`, 并经检查替换后是否还存在`"LB"`:

In [26]:
# 对每个观察值 "production_countries" 列的列表运用上面的函数
Cleaned_Titles['production_countries'] = Cleaned_Titles['production_countries'].replace({"Lebanon": "LB"})

# 检查 "Lebanon" 是否还存在
with pd.option_context('display.max_rows', None):
    print(Cleaned_Titles.explode('production_countries')['production_countries'].value_counts())

production_countries
US    5651
IN    1610
GB    1068
JP    1046
FR     720
ES     637
KR     637
CA     608
DE     383
CN     295
MX     264
IT     224
BR     221
AU     217
TR     195
PH     192
AR     150
ID     149
BE     148
TW     133
NG     131
PL     126
ZA     103
HK     102
NL     102
CO      94
EG      93
DK      89
TH      87
SE      81
LB      71
NO      68
AE      52
IE      49
SG      47
XX      43
IL      42
RU      41
CL      35
CH      33
PS      32
BG      31
MY      30
SA      28
AT      28
IS      28
NZ      27
LU      27
PE      26
RO      25
QA      24
CZ      22
JO      19
FI      18
HU      18
MA      15
UY      15
PT      14
KW      10
KH      10
PK       9
PR       9
UA       8
VN       8
MT       8
SU       7
CD       7
TN       7
LT       7
IR       7
GH       6
SN       6
AL       6
KE       6
IQ       5
MU       5
CY       5
TZ       4
SY       4
MC       4
IO       4
KN       4
GR       4
BD       3
BS       3
DZ       3
GL       3
AO       3
CM       3


另外，里面还存在空字符串表示的国家代码，并非有效数据。但由于出品国家并非所需的关键信息，所以可以保留出品国家为空的观察值。<br>
<br>
针对`Original_Credits`, 不一致数据可能存在于`role`中，我们将查看是否存在多个不同值指代同一演员类型的情况。

In [27]:
Original_Credits['role'].value_counts()

role
ACTOR       73251
DIRECTOR     4550
Name: count, dtype: int64

从以上输出结果来看, `role`只有两种可能的值, `ACTOR`或`DIRECTOR`, 不存在不一致数据。我们可以把这列的类型转换为`Category`, 好处是比字符串类型更节约内存空间，也能表明说值的类型有限。

In [28]:
Cleaned_Credits['role'] = Cleaned_Credits['role'].astype('category')
Cleaned_Credits['role']

0           ACTOR
1           ACTOR
2           ACTOR
3           ACTOR
4           ACTOR
           ...   
77796       ACTOR
77797       ACTOR
77798       ACTOR
77799       ACTOR
77800    DIRECTOR
Name: role, Length: 77801, dtype: category
Categories (2, object): ['ACTOR', 'DIRECTOR']

#### 处理无效或错误数据

可以通过DataFrame的`describe`方法，对数值统计信息进行快速了解。

In [29]:
Original_Titles.describe()

release_year      runtime      seasons   imdb_score    imdb_votes  \
count   5850.000000  5850.000000  2106.000000  5368.000000  5.352000e+03   
mean    2016.417094    76.888889     2.162868     6.510861  2.343938e+04   
std        6.937726    39.002509     2.689041     1.163826  9.582047e+04   
min     1945.000000     0.000000     1.000000     1.500000  5.000000e+00   
25%     2016.000000    44.000000     1.000000     5.800000  5.167500e+02   
50%     2018.000000    83.000000     1.000000     6.600000  2.233500e+03   
75%     2020.000000   104.000000     2.000000     7.300000  9.494000e+03   
max     2022.000000   240.000000    42.000000     9.600000  2.294231e+06   

       tmdb_popularity   tmdb_score  
count      5759.000000  5539.000000  
mean         22.637925     6.829175  
std          81.680263     1.170391  
min           0.009442     0.500000  
25%           2.728500     6.100000  
50%           6.821000     6.900000  
75%          16.590000     7.537500  
max        2274.044000    10.000000

从以上统计信息来看, `Original_Titles`里不存在脱离现实意义的数值。<br>
<br>
`Original_Credits`由于不包含表示数值含义的变量，因此无需用`describe`检查。

#### 整理数据

In [30]:
Cleaned_Titles

id                               title   type  \
1       tm84618                         Taxi Driver  MOVIE   
1       tm84618                         Taxi Driver  MOVIE   
2      tm154986                         Deliverance  MOVIE   
2      tm154986                         Deliverance  MOVIE   
2      tm154986                         Deliverance  MOVIE   
...         ...                                 ...    ...   
5846   tm898842                         C/O Kaadhal  MOVIE   
5847  tm1059008                             Lokillo  MOVIE   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   

                                            description release_year  \
1     A mentally unstable Vietnam War veteran works ...   1976-01-01   
1     A mentally unstable Vietnam War veteran works ...   1976-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
...                                                 ...          ...   
5846  A heart warming film that explores the concept...   2021-01-01   
5847  A controversial TV host and comedian who has b...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   

     age_certification  runtime     genres production_countries  seasons  \
1                    R      114      drama                   US      NaN   
1                    R      114      crime                   US      NaN   
2                    R      109      drama                   US      NaN   
2                    R      109     action                   US      NaN   
2                    R      109   thriller                   US      NaN   
...                ...      ...        ...                  ...      ...   
5846               NaN      134      drama                  NaN      NaN   
5847               NaN       90     comedy                   CO      NaN   
5849               NaN        7     family                  NaN      1.0   
5849               NaN        7  animation                  NaN      1.0   
5849               NaN        7     comedy                  NaN      1.0   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1      tt0075314         8.2    808582.0           40.965       8.179  
1      tt0075314         8.2    808582.0           40.965       8.179  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
...          ...         ...         ...              ...         ...  
5846  tt11803618         7.7       348.0              NaN         NaN  
5847  tt14585902         3.8        68.0           26.005       6.300  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  

[16976 rows x 15 columns]

In [31]:
Cleaned_Credits

person_id         id                name                character  \
0          3748    tm84618      Robert De Niro            Travis Bickle   
1         14658    tm84618        Jodie Foster            Iris Steensma   
2          7064    tm84618       Albert Brooks                      Tom   
3          3739    tm84618       Harvey Keitel  Matthew 'Sport' Higgins   
4         48933    tm84618     Cybill Shepherd                    Betsy   
...         ...        ...                 ...                      ...   
77796    736339  tm1059008    Adelaida Buscato                María Paz   
77797    399499  tm1059008  Luz Stella Luengas             Karen Bayona   
77798    373198  tm1059008         Inés Prieto                    Fanny   
77799    378132  tm1059008        Isabel Gaona                   Cacica   
77800   1950416  tm1059008      Julian Gaviria                      NaN   

           role  
0         ACTOR  
1         ACTOR  
2         ACTOR  
3         ACTOR  
4         ACTOR  
...         ...  
77796     ACTOR  
77797     ACTOR  
77798     ACTOR  
77799     ACTOR  
77800  DIRECTOR  

[77801 rows x 5 columns]

对数据的整理，与分析方向紧密相关。此次数据分析目标是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员，<br>
<br>
那为了能同时获得流派与演员数据，我们需要把`Cleaned_Titles`和`Cleaned_Credits`, 通过`id`作为键进行连接，因为两个数据表中`id`都是影视作品ID。

In [32]:
Credits_With_Titles = pd.merge(Cleaned_Credits, Cleaned_Titles, on = "id", how = "inner")

连接后，我们就能知道各个演职员参与过的影视作品的具体信息。

In [33]:
Credits_With_Titles.head()

person_id       id            name      character   role        title  \
0      3748  tm84618  Robert De Niro  Travis Bickle  ACTOR  Taxi Driver   
1      3748  tm84618  Robert De Niro  Travis Bickle  ACTOR  Taxi Driver   
2     14658  tm84618    Jodie Foster  Iris Steensma  ACTOR  Taxi Driver   
3     14658  tm84618    Jodie Foster  Iris Steensma  ACTOR  Taxi Driver   
4      7064  tm84618   Albert Brooks            Tom  ACTOR  Taxi Driver   

    type                                        description release_year  \
0  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
1  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
2  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
3  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
4  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   

  age_certification  runtime genres production_countries  seasons    imdb_id  \
0                 R      114  drama                   US      NaN  tt0075314   
1                 R      114  crime                   US      NaN  tt0075314   
2                 R      114  drama                   US      NaN  tt0075314   
3                 R      114  crime                   US      NaN  tt0075314   
4                 R      114  drama                   US      NaN  tt0075314   

   imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0         8.2    808582.0           40.965       8.179  
1         8.2    808582.0           40.965       8.179  
2         8.2    808582.0           40.965       8.179  
3         8.2    808582.0           40.965       8.179  
4         8.2    808582.0           40.965       8.179

由于我们只对挖掘演员的参与作品口碑感兴趣，导演不在我们的分析范围内，因此根据`role`, 筛选出类型为`ACTOR`的观察值, 供后续分析。

In [34]:
Actor_With_Titles = Credits_With_Titles.query('role == "ACTOR"')

为了挖掘出各个流派中的高IMDB评分作品演员，我们需要先根据流派和演员进行分组。<br>
<br>
对演员进行分组的时候，选择的是用`person_id`而不是`name`变量，原因是名字容易出现拼错或者重名的情况，演员ID会比演员姓名更加准确的反应是哪位演员。

In [35]:
Groupby_Genres_And_Person_ID = Actor_With_Titles.groupby(["genres", "person_id"])

分组后，我们只需要对`imdb_score`的值进行聚合计算, 因此只提取`imdb_score`变量, 然后调用`mean`, 来计算各个流派影视作品中，每位演员参演作品的平均IMDB评分。

In [36]:
IMDB_Score_Groupby_Genres_And_Person_ID = Groupby_Genres_And_Person_ID["imdb_score"].mean()
IMDB_Score_Groupby_Genres_And_Person_ID

genres   person_id
action   1000         6.866667
         100007       7.000000
         100013       6.400000
         100019       6.500000
         100020       6.500000
                        ...   
western  993735       6.500000
         998673       7.300000
         998674       7.300000
         998675       7.300000
         99940        4.000000
Name: imdb_score, Length: 168881, dtype: float64

我们可以调用`reset_index`, 对层次化索引进行重置, 得到更加规整的DataFrame。

In [37]:
IMDB_Score_Groupby_Genres_And_Person_ID_DF = IMDB_Score_Groupby_Genres_And_Person_ID.reset_index()
IMDB_Score_Groupby_Genres_And_Person_ID_DF

genres person_id  imdb_score
0        action      1000    6.866667
1        action    100007    7.000000
2        action    100013    6.400000
3        action    100019    6.500000
4        action    100020    6.500000
...         ...       ...         ...
168876  western    993735    6.500000
168877  western    998673    7.300000
168878  western    998674    7.300000
168879  western    998675    7.300000
168880  western     99940    4.000000

[168881 rows x 3 columns]

现在针对流派和演员分组的IMDB评分数据已经整理好, 可以进入后续的分析步骤了。<br>
<br>
但我们当前可以继续做一些数据整理，比如对上面的结果再次进行分组，找出各个流派里演员作品最高的平均评分是多少、最高评分对应的演员名字是什么。<br>
<br>
要得到这一结果，我们需要再次用`genres`进行分组，然后取出`imdb_score`变量，计算其最大值。<br>

In [38]:
Genres_Max_Scores = IMDB_Score_Groupby_Genres_And_Person_ID_DF.groupby("genres")["imdb_score"].max()
Genres_Max_Scores

genres
action           9.3
animation        9.3
comedy           9.2
crime            9.5
documentation    9.1
drama            9.5
european         8.9
family           9.3
fantasy          9.3
history          9.1
horror           9.0
music            8.8
reality          8.9
romance          9.2
scifi            9.3
sport            9.1
thriller         9.5
war              8.8
western          8.9
Name: imdb_score, dtype: float64

在我们知道最高分后, 可以把以上结果和之前得到的`IMDB_Score_Groupby_Genres_And_Person_ID_DF`再次进行连接, 得到最高分对应的演员ID是什么，也就是这个最高平均分是哪位演员拿到的。

In [39]:
Genres_Max_Score_With_Person_ID = pd.merge(IMDB_Score_Groupby_Genres_And_Person_ID_DF, Genres_Max_Scores, on = ["genres", "imdb_score"])
Genres_Max_Score_With_Person_ID

genres person_id  imdb_score
0     action     12790         9.3
1     action      1303         9.3
2     action     21033         9.3
3     action    336830         9.3
4     action     86591         9.3
..       ...       ...         ...
131      war    826547         8.8
132  western     22311         8.9
133  western     28166         8.9
134  western     28180         8.9
135  western     93017         8.9

[136 rows x 3 columns]

从以上结果可以看出，最高分对应的演员不一定只有一位，可能有多位演员的平均得分相同。<br>
<br>
为了得到演员ID所对应的演员名字, 我们可以和`Cleaned_Credits`这个DataFrame进行连接，这个DataFrame还有其它列，我们只需要得到`person_id`和`name`的对应, 所以可以先提取出那两列, 并把重复进行删除。

In [40]:
Actor_ID_With_Names = Cleaned_Credits[['person_id', 'name']].drop_duplicates()
Actor_ID_With_Names.head(10)

person_id             name
0      3748   Robert De Niro
1     14658     Jodie Foster
2      7064    Albert Brooks
3      3739    Harvey Keitel
4     48933  Cybill Shepherd
5     32267      Peter Boyle
6    519612   Leonard Harris
7     29068   Diahnne Abbott
8    519613      Gino Ardito
9      3308  Martin Scorsese

下一步就可以把`Actor_ID_With_Names`与前面得到的`Genres_Max_Score_With_Person_ID`进行连接，增加`name`变量, 从而展示平均评分最高的演员名字。

In [41]:
Genres_Max_Score_With_Actor_Name = pd.merge(Genres_Max_Score_With_Person_ID, Actor_ID_With_Names, on = "person_id")
Genres_Max_Score_With_Actor_Name

genres person_id  imdb_score                name
0     action     12790         9.3         Olivia Hack
1     action      1303         9.3       Jessie Flower
2     action     21033         9.3          Zach Tyler
3     action    336830         9.3     André Sogliuzzo
4     action     86591         9.3       Cricket Leigh
..       ...       ...         ...                 ...
131      war    826547         8.8         Yuto Uemura
132  western     22311         8.9     Koichi Yamadera
133  western     28166         8.9  Megumi Hayashibara
134  western     28180         8.9      Unsho Ishizuka
135  western     93017         8.9            Aoi Tada

[136 rows x 4 columns]

In [42]:
Genres_Max_Score_With_Actor_Name = Genres_Max_Score_With_Actor_Name.sort_values("genres").reset_index().drop("index", axis = 1)
Genres_Max_Score_With_Actor_Name

genres person_id  imdb_score                name
0     action     12790         9.3         Olivia Hack
1     action      1303         9.3       Jessie Flower
2     action     21033         9.3          Zach Tyler
3     action    336830         9.3     André Sogliuzzo
4     action     86591         9.3       Cricket Leigh
..       ...       ...         ...                 ...
131      war    826547         8.8         Yuto Uemura
132  western     28180         8.9      Unsho Ishizuka
133  western     22311         8.9     Koichi Yamadera
134  western     28166         8.9  Megumi Hayashibara
135  western     93017         8.9            Aoi Tada

[136 rows x 4 columns]

$$ \oint_{\partial V} \mathbf{E} \cdot d\mathbf{A} = \frac{Q_{\text{enc}}}{\varepsilon_0} $$